File: Challenges with Solutions/is-the-binary-search-tree-balanced-cherelemma-main/challenge.rb
class Node
  attr_reader :data
  attr_accessor :left, :right
  
  def initialize data
    @data = data
  end
end

# method to create a binary tree
def array_to_tree(array, i = 0)
	# use your function from the previous challenge
	return nil if (i >= array.length || array[i] === 0)
	node = Node.new(array[i])
	node.left = array_to_tree(array, 2*i + 1)
	node.right = array_to_tree(array, 2*i + 2)
	return node
end

# helper function: find height of sub-tree
def height(node)
	return 0 if (node.nil?)
	
	# Get the height of the left and right subtree 
	# of the current node using recursion.
	leftHeight = height(node.left)
	rightHeight = height(node.right)
	
	# Choose the larger one and add the root height to it.
	return leftHeight  + 1 if (leftHeight > rightHeight)
	return rightHeight + 1
end

def balanced_tree?(array)
  root = array_to_tree(array)
	
	# call a function to check if the tree is balanced BST or not
  balanced?(root)
end

def balanced? (node)
  return true if node.nil?
  left_height = height(node.left)   # calculate left subtree's height
  right_height = height(node.right)  # calculate right subtree's height
  return false if (left_height - right_height).abs > 1

  # Are the left and right subtree balanced? Check it using recursive
  balanced?(node.left) && balanced?(node.right)
end

puts balanced_tree?([1, 2, 0, 3, 4, 0, 0])
# => false

puts balanced_tree?([1, 2, 3, 4, 5, 6, 7])
# => true

File: Challenges with Solutions/how-tall-is-the-tree-cherelemma-main/challenge.js
class Node {
  constructor(data) {
    this.data = data
    this.left = null
    this.right = null
  }
}

function arrayToTree(array, i = 0) {
  if (i >= array.length || array[i] === 0) {
    return null
  }

  const node = new Node(array[i])
  node.left = arrayToTree(array, 2*i+1)
  node.right = arrayToTree(array, 2*i+2)

  return node
}

// helper function: find height of sub-tree
function height(node) {
	if (!node) return 0;
	
	// Get the height of the left and right subtree of the current node using recursion.
	let leftHeight = height(node.left);
	let rightHeight = height(node.right);
	
	// Choose the larger one and add the root height to it.
	if (leftHeight > rightHeight) return leftHeight  + 1
	else return rightHeight + 1;
}

function binaryTreeHeight(arrayTree) {
  // write your code here
  const tree = arrayToTree(arrayTree);
  return height(tree);
}

console.log(binaryTreeHeight([2, 7, 5, 2, 6, 0, 9]))
// => 3

console.log(binaryTreeHeight([2, 7, 5, 2, 6, 0, 9, 0, 8, 1, 4]))
// => 4

module.exports = binaryTreeHeight
